src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,225 @@
 
     @Override
     public void add(T value, int index) {
+        checkPositionIndex(index);
+        if (index == size) {
+            final Node<T> l = tail;
+            final Node<T> newNode = new Node<>(l, value, null);
+            tail = newNode;
+            if (l == null) {
+                head = newNode;
+            } else {
+                l.next = newNode;
+            }
+            size++;
+        } else {
+            Node<T> previous = head;
+            for (int i = 0; i < index; i++) {
+                previous = previous.next;
+            }
+            final Node<T> pred = previous.prev;
+            final Node<T> newNode = new Node<>(pred, value, previous);
+            previous.prev = newNode;
+            if (pred == null) {
+                head = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        Object[] a = list.toArray();
+        final int numNew = a.length;
+        Node<T> pred;
+        Node<T> succ;
+        succ = null;
+        pred = tail;
+        for (Object o : a) {
+            T e = (T) o;
+            Node<T> newNode = new Node<>(pred, e, null);
+            if (pred == null) {
+                head = newNode;
+            } else {
+                pred.next = newNode;
+            }
+            pred = newNode;
+        }
+
+        if (succ == null) {
+            tail = pred;
+        } else {
+            pred.next = succ;
+            succ.prev = pred;
+        }
+        size += numNew;
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index);
+        return node(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index);
+        Node<T> current = node(index);
+        T finVal = current.value;
+        current.value = value;
+        return finVal;
+    }
+
+    public Node<T> node(int index) {
+        if (index < (size >> 1)) {
+            Node<T> x = head;
+            for (int i = 0; i < index; i++) {
+                x = x.next;
+            }
+            return x;
+        } else {
+            Node<T> x = tail;
+            for (int i = size - 1; i > index; i--) {
+                x = x.prev;
+            }
+            return x;
+        }
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index);
+        if (index == 0) {
+            final T temp = head.value;
+            head = head.next;
+            size--;
+            if (head == null) {
+                tail = null;
+            }
+            return temp;
+        } else if (index == size - 1) {
+            if (size == 0) {
+                return null;
+            } else if (size == 1) {
+                T temp = head.value;
+                head = tail = null;
+                size = 0;
+                return temp;
```

Comment:
redundant block, if size is 1 then you checkIndex will allow only index == 0 and this logic is already mentioned above